frontend/src/components/BackendEndpoint.tsx

```diff
@@ -0,0 +1,17 @@
+function getHostAndPort() {
```

Comment:
This is not a component, its just some functions. It's only a component if it is part of the React tree. Therefore I'd suggest moving it to the top level of the frontend src code. Also every call evaluates all variables. They are known at bundle-time and can be preserved one time for every additional call.

Preferable a generic HTTP/WS interface would be great, using the information here, so that clients can just issue an API-Request and not worry about the hostname, same for the websocket connection.

In the simplest case, returning a partially-built Request object (https://developer.mozilla.org/en-US/docs/Web/API/Request) might be a quick implementation without changing a lot of the existing code:

```
const hostname = "http://hostname"
function buildAPIRequest(pathAndQuery: string, options: any){
   return new Request(`{$hostname}/${pathAndQuery}`, options)
}

function buildAPIRequest(pathAndQuery: string){
  //we can use typescript to force the user to start the path with a /
   return new Request(`{$hostname}${pathAndQuery}`)
}

```
This interface can be optimized by forcing consumers to use the path-slash in their string-arguments: https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/

consumers can still call fetch with the request object or manipulate it as desired, but do not have to worry about the hostname any longer:

```
import {makeAPIRequest} from './somewhere'

const req = makeAPIRequest('/user/login' , {method: 'POST'})
//optionally set body of req, etc...
fetch(req)...

```

Same approach can be used for the WebSocket, giving a path+query as params and returning a WebSocket that uses that path+query with the hostname, hiding it from the client component.

Reply:
Since current TypeScript version used in the repo does not support that new feature, a string as param is sufficient and that one can be checker for a starting slash.